### PR TITLE
Write to Zarr store using h5py-like interface

### PR DESCRIPTION
**Pull request by magland**
This PR adds support for writing to Zarr stores using the h5py-like interface.

See https://github.com/NeurodataWithoutBorders/lindi/blob/write/tests/test_zarr_write.py

and

https://github.com/NeurodataWithoutBorders/lindi/blob/write/examples/example_create_zarr_nwb.py

### CODE REVIEW

**Comment by oruebel** [comment]
Just to clarify, if I understand this correctly, `lindi.LindiH5pyFile` is here backed by a Zarr store, i.e., it is representing a Zarr file as if it where `h5py` and as such, `pynwb.NWBHDF5IO` is here actually writing to Zarr. I.e., in some sense this would be in a form a replacement for`ZarrIO`  in `hdmf_zarr`.

https://github.com/NeurodataWithoutBorders/lindi/blob/0f6814a0a7303ef4ce208cfde2d8980c112fe92d/examples/example_create_zarr_nwb.py#L16-L20

**Comment by magland** [comment]
> Just to clarify, if I understand this correctly, `lindi.LindiH5pyFile` is here backed by a Zarr store, i.e., it is representing a Zarr file as if it where `h5py` and as such, `pynwb.NWBHDF5IO` is here actually writing to Zarr. I.e., in some sense this would be in a form a replacement for`ZarrIO` in `hdmf_zarr`.
> 
> https://github.com/NeurodataWithoutBorders/lindi/blob/0f6814a0a7303ef4ce208cfde2d8980c112fe92d/examples/example_create_zarr_nwb.py#L16-L20

@oruebel Yes that's right.

File: examples/example_edit_nwb.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+# Create a new reference file system
+rfs_new = client.to_reference_file_system()
+
+# Optionally write to a file
+# import json
+# with open('new.zarr.json', 'w') as f:
+#     json.dump(rfs_new, f)
+
+# Load a new h5py-like client from the new reference file system
+client_new = lindi.LindiH5pyFile.from_reference_file_system(rfs_new)
```
Comment (rly) [inline examples/example_edit_nwb.py:27]: Could the same result be achieved by editing the downloaded JSON directly, without going through h5py to delete the dataset and replace the dataset? I imagine this is easier for attributes, the values of which are stored in the JSON, but I imagine this can be done on datasets as well with careful handling. Going through h5py is convenient, but in a way, it adds unnecessary complexity because the data isn't an HDF5 file anymore; it's a JSON + references to chunked, compressed, and/or base64-encoded datasets.

File: tests/test_fletcher32.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        with h5py.File(filename, 'w') as f:
+            dset = f.create_dataset('dset', shape=(100,), dtype='i4', fletcher32=True)
+            dset[...] = range(100)
+            assert dset.fletcher32
+        store = lindi.LindiH5ZarrStore.from_file(filename, url=filename)
+        rfs = store.to_reference_file_system()
+        client = lindi.LindiH5pyFile.from_reference_file_system(rfs)
+        ds0 = client['dset']
+        assert isinstance(ds0, h5py.Dataset)
+        data = ds0[...]
```
Comment (rly) [inline tests/test_fletcher32.py:20]:
```suggestion
        assert ds0.fletcher32
        data = ds0[...]
```

Should we check that the fletcher32 of the read dataset is True?

File: tests/test_remote_data.py [hunk 3]
```diff
@@ -43,3 +44,52 @@ def test_remote_data_2():
...
+    url = 'https://lindi.neurosift.org/dandi/dandisets/000939/assets/11f512ba-5bcf-4230-a8cb-dc8d36db38cb/zarr.json'
+
+    client = lindi.LindiH5pyFile.from_reference_file_system(url)
+
+    rfs2 = {'refs': {
+        '.zgroup': '{"zarr_format": 2}',
+    }}
+    client2 = lindi.LindiH5pyFile.from_reference_file_system(rfs2)
+
+    # This first dataset has is a 2D array with chunks
```
Comment (rly) [inline tests/test_remote_data.py]:
```suggestion
    # This first dataset is a 2D array with chunks
```

File: tests/test_zarr_write.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    for k in h5f2.keys():
+        if k not in h5f:
+            raise Exception(f'Key {k} not found in h5f')
+
+
+def _assert_datasets_equal(h5d1: h5py.Dataset, h5d2: h5py.Dataset):
+    print(f'Comparing datasets: {h5d1.name}')
+    assert h5d1.shape == h5d2.shape, f'h5d1.shape: {h5d1.shape}, h5d2.shape: {h5d2.shape}'
+    assert h5d1.dtype == h5d2.dtype, f'h5d1.dtype: {h5d1.dtype}, h5d2.dtype: {h5d2.dtype}'
+    assert h5d1.dtype == h5d2.dtype, f'h5d1.dtype: {h5d1.dtype}, h5d2.dtype: {h5d2.dtype}'
```
Comment (rly) [inline tests/test_zarr_write.py]:
```suggestion
```

It might also be useful to compare the other important dataset properties like compression, chunks, fill value, fletcher32, etc. but that can be done in a separate PR.

File: examples/example_edit_nwb.py (hunk 1 above)
Comment (magland) [reply examples/example_edit_nwb.py:27]: @rly I actually think going through h5py is the right way to do this because it uses zarr to make all the manipulations.

Comment (rly) [reply examples/example_edit_nwb.py:27]: I see. That makes sense.

File: lindi/LindiH5pyFile/LindiH5pyAttributes.py [hunk 5]
```diff
@@ -46,14 +54,14 @@ def __getitem__(self, key):
...
+            val = decode_nan_inf_ninf(val)
 
-            return val
+            return zarr_to_h5_attr(val)
         else:
             raise ValueError(f"Unknown attrs_type: {self._attrs_type}")
 
     def __setitem__(self, key, value):
-        raise KeyError("Cannot set attributes on read-only object")
+        self._write.__setitem__(key, value)
```
Comment (rly) [inline lindi/LindiH5pyFile/LindiH5pyAttributes.py]: This is more of a stylistic suggestion than functional. I think it would be more intuitive to have `__init__` be:
```python
    def __init__(self, attrs, attrs_type: Literal["h5py", "zarr"], readonly: bool):
        self._attrs = attrs
        self._attrs_type = attrs_type
        self._readonly = readonly
        if readonly:
            self._writer = LindiH5pyAttributesWrite(self)
```
and this function to be:
```python
    def __setitem__(self, key, value):
        if self._readonly:
            raise KeyError("Cannot set attributes on read-only object")
        self._writer.__setitem__(key, value)
```
(Move the readonly check here and name the object `writer` instead of `write`.)
This mirrors the readonly check in `LindiH5pyDataset.__setitem__` and `LindiH5pyGroup.__setitem__`.

File: lindi/LindiH5pyFile/LindiH5pyDataset.py [hunk 6]
```diff
@@ -34,9 +36,19 @@ def __init__(self, _dataset_object: Union[h5py.Dataset, zarr.Array], _file: "Lin
         if isinstance(_dataset_object, zarr.Array):
             compound_dtype_obj = _dataset_object.attrs.get("_COMPOUND_DTYPE", None)
             if compound_dtype_obj is not None:
+                assert isinstance(compound_dtype_obj, list)
+                for i in range(len(compound_dtype_obj)):
```
Comment (rly) [inline lindi/LindiH5pyFile/LindiH5pyDataset.py:43]:
```suggestion
                # compound_dtype_obj is a list of tuples (name, dtype)
                # where dtype == "<REFERENCE>" if it represents an HDF5 reference
                for i in range(len(compound_dtype_obj)):
```

Adding comments as a reminder of the internal format

File: lindi/LindiH5pyFile/LindiH5pyFile.py [hunk 7]
```diff
@@ -14,19 +14,35 @@
...
         """
         Create a LindiH5pyFile from a reference file system.
+
+        Parameters
+        ----------
+        rfs : Union[dict, str]
+            The reference file system. This can be a dictionary or a URL or path
+            to a .zarr.json file.
+        mode : Literal["r", "r+"], optional
+            The mode to open the file object in, by default "r" If the mode is
```
Comment (rly) [inline lindi/LindiH5pyFile/LindiH5pyFile.py]:
```suggestion
            The mode to open the file object in (default: "r"). If the mode is
```

File: lindi/LindiH5pyFile/LindiH5pyFile.py [hunk 8]
```diff
@@ -14,19 +14,35 @@
...
+            The reference file system. This can be a dictionary or a URL or path
+            to a .zarr.json file.
+        mode : Literal["r", "r+"], optional
+            The mode to open the file object in (default: "r"). If the mode is
+            "r", the file object will be read-only. If the mode is "r+", the
+            file will be read-write. However, if the rfs is a string (URL or
+            path), the file itself will not be modified on changes, but the
+            internal in-memory representation will be modified. Use
+            to_reference_file_system() to export the updated reference file
+            system.
```
Comment (rly) [inline lindi/LindiH5pyFile/LindiH5pyFile.py]:
```suggestion
            system to the same file or a new file.
```

Comment (rly) [reply lindi/LindiH5pyFile/LindiH5pyFile.py]: > the file itself will not be modified on changes

This does not seem intuitive. We can merge but I'll think about what might be a more intuitive interface for users.

File: lindi/LindiH5pyFile/LindiH5pyFile.py [hunk 9]
```diff
@@ -14,19 +14,35 @@
...
         """
         Create a LindiH5pyFile from a reference file system.
+
+        Parameters
+        ----------
+        rfs : Union[dict, str]
+            The reference file system. This can be a dictionary or a URL or path
+            to a .zarr.json file.
+        mode : Literal["r", "r+"], optional
+            The mode to open the file object in (default: "r"). If the mode is
```
Comment (rly) [inline lindi/LindiH5pyFile/LindiH5pyFile.py]:
```suggestion
            The mode to open the file object in, by default "r". If the mode is
```

File: lindi/LindiH5pyFile/LindiH5pyFile.py [hunk 10]
```diff
@@ -14,19 +14,35 @@
...
         self._file_object = _file_object
+        self._zarr_store = _zarr_store
+        self._mode: Literal['r', 'r+'] = _mode
         self._the_group = LindiH5pyGroup(_file_object, self)
 
     @staticmethod
-    def from_reference_file_system(rfs: Union[dict, str]):
+    def from_reference_file_system(rfs: Union[dict, str], mode: Literal["r", "r+"] = "r"):
         """
         Create a LindiH5pyFile from a reference file system.
```
Comment (rly) [inline lindi/LindiH5pyFile/LindiH5pyFile.py:31]: Would any RFS JSON, e.g., one generated by `kerchunk` work here? It should be fine, right?

File: lindi/LindiH5pyFile/LindiH5pyFile.py [hunk 11]
```diff
@@ -137,9 +191,54 @@ def __enter__(self):  # type: ignore
...
+        if isinstance(self._file_object, h5py.File):
+            return self._file_object.__hash__()
+        else:
+            return id(self)
+
+    def copy(self, source, dest, name=None,
+             shallow=False, expand_soft=False, expand_external=False,
+             expand_refs=False, without_attrs=False):
+        if shallow:
+            raise Exception("shalle is not implemented for copy")
```
Comment (rly) [inline lindi/LindiH5pyFile/LindiH5pyFile.py]:
```suggestion
            raise Exception("shallow is not implemented for copy")
```

File: lindi/LindiH5pyFile/LindiH5pyFile.py [hunk 12]
```diff
@@ -209,6 +308,29 @@ def file(self):
...
+    def ref(self):
+        return self._the_group.ref
+
+    ##############################
+    # write
+    def create_group(self, name, track_order=None):
+        if self._mode not in ['r+']:
+            raise Exception("Cannot create group in read-only mode")
+        if track_order is not None:
+            raise Exception("track_order is not supported (I don't know what it is)")
```
Comment (rly) [inline lindi/LindiH5pyFile/LindiH5pyFile.py:321]: if `track_order` is true when creating an `h5py.Group`, then when iterating through the objects within the group, they are returned in order of creation, otherwise it is in ascending alphanumeric order. PyNWB doesn't use this. But I don't think it would hurt to pass this argument to `self._the_group.create_group(name)` below.

File: lindi/LindiH5pyFile/LindiH5pyFile.py [hunk 13]
```diff
@@ -218,3 +340,60 @@ def _download_file(url: str, filename: str) -> None:
...
+                        if src_ref_key.startswith(f'{src_item_name}/'):
+                            dst_ref_key = f'{name}/{src_ref_key[len(src_item_name) + 1:]}'
+                            # Even though it's not expected to be a problem, we
+                            # do a deep copy here because a problem resulting
+                            # from one rfs being modified affecting another
+                            # would be very difficult to debug.
+                            dst_rfs['refs'][dst_ref_key] = _deep_copy(src_rfs['refs'][src_ref_key])
+                    return
+
+        dst_item = dest.create_dataset(name, data=src_item[()])
```
Comment (rly) [inline lindi/LindiH5pyFile/LindiH5pyFile.py:383]:
```suggestion
        src_item.parent.copy(src_item, dest, name)
```
This should copy all the IO settings and I believe it does not load the dataset into memory.

Comment (rly) [reply lindi/LindiH5pyFile/LindiH5pyFile.py:383]: It will also copy attributes by default, so the two lines below can be removed.

If the dataset is a dataset of references, I'm not sure what will happen.

File: lindi/LindiH5pyFile/LindiH5pyGroup.py [hunk 14]
```diff
@@ -49,12 +53,12 @@ def __getitem__(self, name):
                 soft_link = x.attrs.get('_SOFT_LINK', None)
                 if soft_link is not None:
                     link_path = soft_link['path']
-                    target_grp = self._file.get(link_path)
-                    if not isinstance(target_grp, LindiH5pyGroup):
+                    target_item = self._file.get(link_path)
+                    if not isinstance(target_item, LindiH5pyGroup) and not isinstance(target_item, LindiH5pyDataset):
```
Comment (rly) [inline lindi/LindiH5pyFile/LindiH5pyGroup.py]:
```suggestion
                    if not isinstance(target_item, (LindiH5pyGroup, LindiH5pyDataset)):
```

File: lindi/LindiH5pyFile/LindiH5pyFile.py [hunk 15]
```diff
@@ -14,19 +14,35 @@
 
 
 class LindiH5pyFile(h5py.File):
-    def __init__(self, _file_object: Union[h5py.File, zarr.Group]):
+    def __init__(self, _file_object: Union[h5py.File, zarr.Group], *, _zarr_store: Union[ZarrStore, None] = None, _mode: Literal["r", "r+"] = "r"):
```
Comment (rly) [inline lindi/LindiH5pyFile/LindiH5pyFile.py:17]: In h5py, `h5py.File` is a subclass of `h5py.Group`. Multiple inheritance is ugly but there is some code shared between `LindiH5pyFile` and `LindiH5pyGroup` that *I think* could be simplified by having `LindiH5pyFIle` inherit also from `LindiH5pyGroup`.

File: lindi/conversion/h5_ref_to_zarr_attr.py [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+    # The get_name call can actually be quite slow. A possible way around this
+    # is to do an initial pass through the file and build a map of object IDs to
+    # paths. This would need to happen elsewhere in the code.
+    deref_objname = h5py.h5r.get_name(ref, file_id)
+    if deref_objname is None:
+        raise ValueError(f"Could not dereference object with reference {ref}")
+    deref_objname = deref_objname.decode("utf-8")
+
+    dref_obj = h5f[deref_objname]
```
Comment (rly) [inline lindi/conversion/h5_ref_to_zarr_attr.py]: Do we need to do this, or could we just do `h5f[ref]`? Or is the idea that this code is what h5py does under the hood, we can make it more efficient by later caching `get_name` calls, and exposing this code now makes it easier to optimize later?

File: lindi/conversion/h5_ref_to_zarr_attr.py [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+
+    # Here we assume that the file has a top-level attribute called "object_id".
+    # This will be the case for files created by the LindiH5ZarrStore class.
+    file_object_id = h5f.attrs.get("object_id", None)
+
+    # See https://hdmf-zarr.readthedocs.io/en/latest/storage.html#storing-object-references-in-attributes
+    value = {
+        "object_id": object_id,
+        "path": deref_objname,
+        "source": ".",  # Are we always going to use the top-level object as the source?
```
Comment (rly) [inline lindi/conversion/h5_ref_to_zarr_attr.py:50]: @oruebel Under what circumstances would one want to have `source` not be the top level object and have paths be relative to that?

File: tests/test_zarr_write.py [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+
+
+def compare_example_h5_data(h5f: h5py.File, tmpdir: str):
+    with h5py.File(f'{tmpdir}/for_comparison.h5', 'w') as h5f2:
+        write_example_h5_data(h5f2)
+    with h5py.File(f'{tmpdir}/for_comparison.h5', 'r') as h5f2:
+        _assert_groups_equal(h5f, h5f2)
+
+
+def _assert_groups_equal(h5f: h5py.Group, h5f2: h5py.Group):
```
Comment (rly) [inline tests/test_zarr_write.py]: I suggest moving this and all the methods below to a module `tests/utils.py`, removing the leading underscore, and importing them into the tests that use them. Currently `_arrays_are_equal` is copied to `test_copy.py` and `test_core.py`

File: tests/test_copy.py [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+            with pytest.raises(Exception):
+                # This one is not expected to work. Would be difficult to
+                # implement since this involves low-level operations on
+                # LindiH5pyFile.
+                h5f.copy("X", h5f_2, "Z")
+            h5f_2.copy("X", h5f_2, "Z")
+            assert "Z" in h5f_2
+            assert h5f_2["Z"].attrs['attr1'] == 'value1'  # type: ignore
+            assert "A" in h5f_2["Z"]  # type: ignore
+            assert _arrays_are_equal(h5f["X/A"][()], h5f_2["Z/A"][()])  # type: ignore
```
Comment (rly) [inline tests/test_copy.py]: You could also use:
```
_assert_groups_equal(h5f["X"], h5f_2["Z"])
```
after refactoring test functions to a util function. Similarly below

File: examples/example_create_zarr_nwb.py [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+def example_create_zarr_nwb():
+    zarr_dirname = 'example_nwb.zarr'
+    if os.path.exists(zarr_dirname):
+        shutil.rmtree(zarr_dirname)
+
+    nwbfile = _create_sample_nwb_file()
+
+    store = zarr.DirectoryStore(zarr_dirname)
+    zarr.group(store=store)  # create a root group
+    with lindi.LindiH5pyFile.from_zarr_store(store, mode='a') as client:
```
Comment (rly) [inline examples/example_create_zarr_nwb.py]:
```suggestion
    with lindi.LindiH5pyFile.from_zarr_store(store, mode='r+') as client:
```

`"a"` mode is not implemented below yet

File: lindi/conversion/attr_conversion.py [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+        list_dtype = _determine_list_dtype(attr)
+        attr = np.array(attr, dtype=list_dtype)
+
+    if isinstance(attr, str) and attr in special_strings:
+        raise ValueError(f"Special string {attr} not allowed in attribute value at {label}")
+    if attr is None:
+        raise Exception(f"Unexpected h5 attribute: None at {label}")
+    elif type(attr) in [int, np.int8, np.int16, np.int32, np.int64, np.uint8, np.uint16, np.uint32, np.uint64]:
+        return int(attr)
+    elif type(attr) in [float, np.float16, np.float32, np.float64, np.float128]:
```
Comment (rly) [inline lindi/conversion/attr_conversion.py]: `np.float128` does not exist on my Mac M1 system... `AttributeError: module 'numpy' has no attribute 'float128'. Did you mean: 'float16'?`

Comment (rly) [inline lindi/conversion/attr_conversion.py]:
```suggestion
    elif isinstance(attr, float) or (isinstance(attr, np.ndarray) and np.issubdtype(attr.dtype, np.floating)):
```

File: lindi/conversion/attr_conversion.py [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+        return encode_nan_inf_ninf(float(attr))
+    elif type(attr) in [complex, np.complex64, np.complex128, np.complex256]:
```
Comment (rly) [inline lindi/conversion/attr_conversion.py]:
```suggestion
    elif isinstance(attr, complex) or (isinstance(attr, np.ndarray) and np.issubdtype(attr.dtype, np.complexfloating)):
```

`np.complex256` also doesn't exist on my system

File: lindi/conversion/nan_inf_ninf.py [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+    else:
+        return val
+
+
+def encode_nan_inf_ninf(val):
+    if isinstance(val, list):
+        return [encode_nan_inf_ninf(v) for v in val]
+    elif isinstance(val, dict):
+        return {k: encode_nan_inf_ninf(v) for k, v in val.items()}
+    elif type(val) in [float, np.float16, np.float32, np.float64, np.float128]:
```
Comment (rly) [inline lindi/conversion/nan_inf_ninf.py]:
```suggestion
    elif isinstance(val, float) or (isinstance(val, np.ndarray) and np.issubdtype(val.dtype, np.floating)):
```

**Comment by rly** [comment]
The suggested changes around `np.float128`/`np.complex256` and the change of `mode="a"` to `mode="r+"` were necessary to make `examples/example_create_zarr_nwb.py` run on my machine.

File: lindi/LindiH5pyFile/write/LindiH5pyGroupWrite.py [hunk 24]
```diff
@@ -0,0 +1,110 @@
...
+
+    def require_group(self, name):
+        if name in self.p:
+            ret = self.p[name]
+            if not isinstance(ret, LindiH5pyGroup):
+                raise Exception(f'Expected a group at {name} but got {type(ret)}')
+            return ret
+        return self.create_group(name)
+
+    def create_dataset(self, name, shape=None, dtype=None, data=None, **kwds):
```
Comment (rly) [inline lindi/LindiH5pyFile/write/LindiH5pyGroupWrite.py]: There are a few places in HDMF that call `group.require_dataset` when writing datasets containing references, e.g., https://github.com/hdmf-dev/hdmf/blob/244d17a28ed436849b1973a3aaac8522d0ea922b/src/hdmf/backends/hdf5/h5tools.py#L1185

I do not know why `require_dataset` is called instead of `create_dataset` but that's what it is right now...

It looks like the source code for `Group.require_dataset` returns an `h5py.Dataset` if the dataset exists already. https://github.com/h5py/h5py/blob/066cad17312a592b26f59f91ef383dbde29bd142/h5py/_hl/group.py#L281-L282

Should we return an `LindiH5pyDataset` instead?

**Comment by rly** [comment]
I did not do heavy testing of this PR, but I read through it all, and as far as I can tell, it makes sense and looks good. Nicely done! 

HDMF supports a lot of weird edge cases. I'll see if I can think up more tests to what you already have here.

File: lindi/LindiH5pyFile/LindiH5pyAttributes.py (hunk 5 above)
Comment (magland) [reply lindi/LindiH5pyFile/LindiH5pyAttributes.py]: Thanks. I have made that change throughout.

File: lindi/LindiH5pyFile/LindiH5pyFile.py (hunk 8 above)
Comment (magland) [reply lindi/LindiH5pyFile/LindiH5pyFile.py]: I agree, we'll need to think about it.

File: lindi/LindiH5pyFile/LindiH5pyFile.py (hunk 10 above)
Comment (magland) [reply lindi/LindiH5pyFile/LindiH5pyFile.py:31]: yes it would.

File: lindi/LindiH5pyFile/LindiH5pyFile.py (hunk 12 above)
Comment (magland) [reply lindi/LindiH5pyFile/LindiH5pyFile.py:321]: I'll open a separate issue for this.

File: lindi/LindiH5pyFile/LindiH5pyFile.py (hunk 15 above)
Comment (magland) [reply lindi/LindiH5pyFile/LindiH5pyFile.py:17]: I'd rather duplicate the code than to add that complexity. But it may be important to have `isinstance(lindi_file, h5py.Group)` return True. So I'll open this as a separate issue.

File: lindi/conversion/h5_ref_to_zarr_attr.py (hunk 16 above)
Comment (magland) [reply lindi/conversion/h5_ref_to_zarr_attr.py]: Thanks, that's a big simplification. Not sure why I was doing it the way I was. Made the change.

File: lindi/conversion/h5_ref_to_zarr_attr.py (hunk 17 above)
Comment (magland) [reply lindi/conversion/h5_ref_to_zarr_attr.py:50]: I believe it could be the path of an external file.

File: tests/test_zarr_write.py (hunk 18 above)
Comment (magland) [reply tests/test_zarr_write.py]: done

File: tests/test_copy.py (hunk 19 above)
Comment (magland) [reply tests/test_copy.py]: done

File: examples/example_create_zarr_nwb.py (hunk 20 above)
Comment (magland) [reply examples/example_create_zarr_nwb.py]: fixed

File: lindi/conversion/attr_conversion.py (hunk 21 above)
Comment (magland) [reply lindi/conversion/attr_conversion.py]: What you suggest doesn't quite do the same thing. I'm leaving as is and removing float128

File: lindi/LindiH5pyFile/write/LindiH5pyGroupWrite.py (hunk 24 above)
Comment (magland) [reply lindi/LindiH5pyFile/write/LindiH5pyGroupWrite.py]: Thanks. It was using the `require_dataset()` from the base class, which is not what we want. I have now implemented require_dataset on the LindiH5pyGroup, added a test, and pushed the update.

File: lindi/LindiH5pyFile/LindiH5pyFile.py (hunk 13 above)
Comment (magland) [reply lindi/LindiH5pyFile/LindiH5pyFile.py:383]: I am skipping this change since I'm concerned about creating an infinite recursion. If you feel strongly let me know.

**Comment by magland** [comment]
> I did not do heavy testing of this PR, but I read through it all, and as far as I can tell, it makes sense and looks good. Nicely done!
> 
> HDMF supports a lot of weird edge cases. I'll see if I can think up more tests to what you already have here.

Thanks Ryan!

File: lindi/LindiH5pyFile/write/LindiH5pyGroupWrite.py (hunk 24 above)
Comment (rly) [reply lindi/LindiH5pyFile/write/LindiH5pyGroupWrite.py]: I don't see `require_dataset` anywhere. Where did you add it?

File: lindi/conversion/attr_conversion.py (hunk 21 above)
Comment (rly) [reply lindi/conversion/attr_conversion.py]: Oh - whoops. I see now that this was for scalars. What about `isinstance(attr, (float, np.floating))`?

I still see `np.float128`?

File: lindi/LindiH5pyFile/write/LindiH5pyGroupWrite.py (hunk 24 above)
Comment (rly) [reply lindi/LindiH5pyFile/write/LindiH5pyGroupWrite.py]: Never mind - page did not refresh

File: lindi/conversion/attr_conversion.py (hunk 21 above)
Comment (rly) [reply lindi/conversion/attr_conversion.py]: > I still see np.float128?

Never mind - page did not refresh.

Comment (rly) [reply lindi/conversion/attr_conversion.py]: See minor suggestion above still though

File: lindi/conversion/attr_conversion.py [hunk 25]
```diff
@@ -0,0 +1,162 @@
...
+    if isinstance(attr, str) and attr in special_strings:
+        raise ValueError(f"Special string {attr} not allowed in attribute value at {label}")
+    if attr is None:
+        raise Exception(f"Unexpected h5 attribute: None at {label}")
+    elif type(attr) in [int, np.int8, np.int16, np.int32, np.int64, np.uint8, np.uint16, np.uint32, np.uint64]:
+        return int(attr)
+    elif type(attr) in [float, np.float16, np.float32, np.float64]:
+        return encode_nan_inf_ninf(float(attr))
+    elif isinstance(attr, complex) or (isinstance(attr, np.ndarray) and np.issubdtype(attr.dtype, np.complexfloating)):
+        raise Exception(f"Complex number is not supported at {label}")
```
Comment (rly) [inline lindi/conversion/attr_conversion.py]:
```suggestion
    elif isinstance(attr, (complex, np.complexfloating)):
        raise Exception(f"Complex number is not supported at {label}")
```
Fix on my suggestion

File: lindi/conversion/attr_conversion.py (hunk 21 above)
Comment (magland) [reply lindi/conversion/attr_conversion.py]: Got it. Made the change.

File: lindi/conversion/attr_conversion.py (hunk 25 above)
Comment (magland) [reply lindi/conversion/attr_conversion.py]: made the change.

File: lindi/LindiH5pyFile/LindiH5pyFile.py (hunk 13 above)
Comment (rly) [reply lindi/LindiH5pyFile/LindiH5pyFile.py:383]: OK, that's fine

**Review by rly — APPROVED** [review]
